cmd/compute-domain-controller/daemonset.go

```diff
@@ -0,0 +1,398 @@
+	podManagers                  map[string]*DaemonSetPodManager
+}
+
+func NewDaemonSetManager(config *ManagerConfig, getComputeDomain GetComputeDomainFunc) *DaemonSetManager {
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		config.clientsets.Core,
+		informerResyncPeriod,
+		informers.WithNamespace(config.driverNamespace),
+		informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+			opts.LabelSelector = metav1.FormatLabelSelector(labelSelector)
+		}),
+	)
+
+	informer := factory.Apps().V1().DaemonSets().Informer()
+
+	m := &DaemonSetManager{
+		config:           config,
+		getComputeDomain: getComputeDomain,
+		factory:          factory,
+		informer:         informer,
+		podManagers:      make(map[string]*DaemonSetPodManager),
+	}
+	m.resourceClaimTemplateManager = NewDaemonSetResourceClaimTemplateManager(config)
+
+	return m
+}
+
+func (m *DaemonSetManager) Start(ctx context.Context) (rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancelContext = cancel
+
+	defer func() {
+		if rerr != nil {
+			if err := m.Stop(); err != nil {
+				klog.Errorf("error stopping DaemonSet manager: %v", err)
+			}
+		}
+	}()
+
+	if err := addComputeDomainLabelIndexer[*appsv1.DaemonSet](m.informer); err != nil {
+		return fmt.Errorf("error adding indexer for MulitNodeEnvironment label: %w", err)
+	}
+
+	_, err := m.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			m.config.workQueue.Enqueue(obj, m.onAddOrUpdate)
+		},
+		UpdateFunc: func(objOld, objNew any) {
+			m.config.workQueue.Enqueue(objNew, m.onAddOrUpdate)
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error adding event handlers for DaemonSet informer: %w", err)
+	}
+
+	m.waitGroup.Add(1)
+	go func() {
+		defer m.waitGroup.Done()
+		m.factory.Start(ctx.Done())
+	}()
+
+	if !cache.WaitForCacheSync(ctx.Done(), m.informer.HasSynced) {
+		return fmt.Errorf("informer cache sync for DaemonSet failed")
+	}
+
+	if err := m.resourceClaimTemplateManager.Start(ctx); err != nil {
+		return fmt.Errorf("error starting ResourceClaimTemplate manager: %w", err)
+	}
+
+	return nil
+}
+
+func (m *DaemonSetManager) Stop() error {
+	if err := m.removeAllPodManagers(); err != nil {
+		return fmt.Errorf("error removing all Pod managers: %w", err)
+	}
+	m.cancelContext()
+	m.waitGroup.Wait()
+	return nil
+}
+
+func (m *DaemonSetManager) Create(ctx context.Context, namespace string, cd *nvapi.ComputeDomain) (*appsv1.DaemonSet, error) {
+	ds, err := getByComputeDomainUID[*appsv1.DaemonSet](ctx, m.informer, string(cd.UID))
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving DaemonSet: %w", err)
+	}
+	if len(ds) > 1 {
+		return nil, fmt.Errorf("more than one DaemonSet found with same ComputeDomain UID")
```

Comment:
:+1: I like the defensive invariant checking. And I used the opportunity to read https://go.dev/doc/faq#assertions.